File-Backed Genome::Config Implementation

nnutter opened the pull request:
I still don't consider this ready for merge but I think this is ready for a stringent code review.  After any issues with this pass of code review are addressed I'll start looking at deploying specs and configs and converting the code base to using them.

There are three main classes:

- `Genome::Config`
- `Genome::ConfigSpec`
- `Genome::Configurable`

Each has POD describing them so I won't repeat that here but `Genome::Config` is sort of the base-level API to lookup values and `Genome::Configurable` is a higher-level role to integrate `Genome::Config` into a class definition, typically a Command.

To continue to try and keep the PR reviewable I'm separating out the conversion of the codebase as a separate PR, see nnutter/genome#2.

tmooney commented:
This reminds me somewhat of [genome/configuration-manager](https://github.com/genome/configuration-manager).

davidlmorton commented:
I have a couple of concerns, which you've probably already thought about but I just wanted to make sure they were discussed:

1.  These values are only loaded from files at runtime, so in the case of a Genome::Model::Build, the values could be different for different steps if the configuration changes while a build is running.  This is not ideal, especially with one of the configuration files living in the user's HOME directory.  A user could accidentally crash all their builds by editing that file, and HOME directory slowness would start equating to production slowness.

2.  These values are not persisted in any way, so we lose reproducibility and the ability to debug after the fact.  I know that this is already the case for all of the environment variables in our system currently, so this isn't a regression.  One solution would be to have the Genome::Model::Build framework and maybe the Genome::Process framework concretize the configuration into env variables when a Build or Process is launched, and persist the environment either in a file or in the DB.

acoffman commented:
Is this intended to be general purpose or specific to Genome/UR? If its the later, I'm a little unsure about the idea of the validators. It seems like properties should already have type annotations and classes should validate themselves. Adding the validations to the config seems like we're just duplicating that information in another place.

acoffman commented on lib/perl/Genome/ConfigSpec.pm:
```diff
@@ -0,0 +1,54 @@
+    my $class = shift;
+    my $file = Path::Class::File->new(shift);
+
+    (my $key = $file->basename) =~ s/\.yaml$//i;
```
My gut instinct was that surely `Path::Class::File` had methods for working with file extensions and this manual munging wouldn't be needed, but after looking at the docs, that doesn't seem to be the case. Seems like an odd omission.

brummett replied:
File::Basename::fileparse() will do it, though not case-insensitively.

acoffman replied:
Ah, good to know, thanks for the pointer!

ebelter replied:
You can pass fileparse a suffix list. Presumably you are only looking for '.YAML' and '.yaml'.

nnutter replied:
I switched to using `fileparse`.

acoffman commented on lib/perl/Genome/ConfigSpec.pm:
```diff
@@ -0,0 +1,54 @@
+
+    my @validators;
+    for my $v (@{$data->{validators}}) {
+        my $module_name = join('::', 'Genome', 'ConfigValidator', ucfirst($v));
```
Is it a convention that validator will need to have single word names? (Its fine if so!) This will break if something needed to be camel cased.

nnutter replied:
We can change it as needed.

acoffman commented on lib/perl/Genome/Config.pm:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    for my $v (@{$spec->validators}) {
+        if (my $error = $v->($value, $spec)) {
```
It seems odd that a call to validator's validate method (well, coderef) returns a truthy value on failure and a falsey value on success. I do like how you're using it to sort of auto-generate error messages though.

I wonder if part of the interface of a validator should be an `error_message` class method or something? I also wonder if its worth running all validators rather than dying after the first failure? That way you don't fix one problem only to stumble into the next one when you try again. Obviously this is just pseudocode but maybe something like: 

````perl
my @errors = map { $_->error_message($value) } 
        grep { ! $_->is_valid($value) }
        $spec->validators;

croakf(join(@error_messages, "\n")) if @error_messages;
````

Unrelated, but writing that made me wish perl had fold/reduce in core instead of an external module.

nnutter replied:
I don't think I've quite addressed what you're suggesting but I did make it so the spec class creates the full error message,

```
my @errors = $spec->validate($value);
if (@errors) {
  my $msg = $spec->validation_error(@errors);
  croakf($msg);
}
```

nnutter commented:
I still don't consider this ready for merge but I think this is ready for a stringent code review.  After any issues with this pass of code review are addressed I'll start looking at deploying specs and configs and converting the code base to using them.

nnutter commented:
@davidlmorton,

>     These values are only loaded from files at runtime, so in the case of a Genome::Model::Build, the values could be different for different steps if the configuration changes while a build is running. This is not ideal, especially with one of the configuration files living in the user's HOME directory. A user could accidentally crash all their builds by editing that file, and HOME directory slowness would start equating to production slowness.

I'm not too concerned about this.  Either jobs should run in controlled environments, e.g. `apipe-builder`, or the user should be responsible for changes they make.  I did add the `sticky` option to "preserve" the value in the (shell) environment but I think most of the current configuration values would not use this as they are not configuration that affect build results.

>     These values are not persisted in any way, so we lose reproducibility and the ability to debug after the fact. I know that this is already the case for all of the environment variables in our system currently, so this isn't a regression. One solution would be to have the Genome::Model::Build framework and maybe the Genome::Process framework concretize the configuration into env variables when a Build or Process is launched, and persist the environment either in a file or in the DB.

I think this is orthogonal to how config values are resolved so long as there is a way to get all config values which there is.

tmooney commented on lib/perl/Genome/ConfigSpec.t:
```diff
@@ -0,0 +1,120 @@
+use strict;
+use warnings;
+
+use Genome;
```
This one doesn't have the `qw()` that the other instances of `use Genome` in this PR do.  (It's a strange world if `Genome` starts exporting subroutines by default!)

nnutter replied:
It's strange world when anything can export subroutines by default!

tmooney commented on lib/perl/Genome/Config.pm:
```diff
@@ -0,0 +1,158 @@
+=item get()
+
+C<get($key)> retrieves the configuration value specified by the input key.  It
+does a hierarchival search for the value,
```
`hierarchival` should be `hierarchical`.  This sentence might be clearer with a clause introducing the list, e.g. `...search for the value, which examines the following places:`.

tmooney commented on lib/perl/Genome/Config.pm:
```diff
@@ -0,0 +1,158 @@
+- environment variable
+- user config file
+- snapshot config file
+- global config files
```
Is the global configuration special in that there are multiple files, but the other locations have only one?

nnutter replied:
Yes, the global directories are specified by a colon-delimited environment variable, `XGENOME_CONFIG_DIRS`, defaulting to `/etc`.

tmooney commented on lib/perl/Genome/Config.pm:
```diff
@@ -0,0 +1,158 @@
+C</etc/genome/foo.yaml>.
+
+If you want the configuration value to be optional you must specify a
+C<default_value> even if it is zero or an empty string but it must be defined.
```
This is a run-on sentence.  One possibility is to split it in two: `...specify a C<default_value>.  The default can be zero or an empty string, but it must be defined.`

tmooney commented on lib/perl/Genome/Config.pm:
```diff
@@ -0,0 +1,158 @@
+        croakf('required config key missing: %s', $key);
+    }
+
+    if (!defined($value)) {
```
Would the logic here be more clear if both this and the previous conditional were combined with a check on `$spec->has_default_value` inside?

tmooney commented on lib/perl/Genome/Config.pm:
```diff
@@ -0,0 +1,158 @@
+    }
+
+    if ($spec->sticky) {
+        unless ($spec->has_env) {
```
Would it make sense for the `ConfigSpec` object to check that `env` is set when `sticky` is set?

nnutter replied:
I'll have to think about that.

tmooney commented on lib/perl/Genome/Config/Command/Get.pm:
```diff
@@ -0,0 +1,26 @@
+use Genome::Config qw();
+
+class Genome::Config::Command::Get {
+    is => 'Command',
```
Is there a reason to prefer `Command` over `Command::V2`?  (It seems like we should stop having two versions, but...)

nnutter replied:
For some reason I thought `Command` was the latest version not `V1`.  I'll switch it.

tmooney commented:
Would it be useful to have a command that is sort of a cross between `genome config get` and `genome config validate` that lists all of the configuration sort of like `git config -l`?

nnutter commented:
> Would it be useful to have a command that is sort of a cross between `genome config get` and `genome config validate` that lists all of the configuration sort of like `git config -l`?

I've thought about it but I'm not sure yet.  I'd probably just make `genome config get` default to all keys unless one or more is specified.

nnutter commented:
One thing I'm worrying about is whether requiring a defined value is practical or not.  I'd like it to be.

tmooney commented on lib/perl/Genome/Config.pm:
```diff
@@ -0,0 +1,158 @@
+
+    my @errors = $spec->validate($value);
+    if (@errors) {
+        my $msg = $spec->validation_error(@errors);
```
Maybe `$spec` could provide a method that's basically
```perl
sub {
  my ($self, $value) = @_;
  my @errors = $self->validate($value);
  return unless @errors;
  return $self->validation_error(@errors); 
}
```

tmooney commented on lib/perl/Genome/Config.pm:
```diff
@@ -0,0 +1,158 @@
+or via the corresponding environment variable.
+
+Configuration values are "registered" by creating a spec file, see
+L<Config::Spec>.  Spec files can go in either the snapshot config directory or
```
This is currently `Genome::ConfigSpec` rather than `Config::Spec`.

tmooney commented on lib/perl/Genome/ConfigSpec.pm:
```diff
@@ -0,0 +1,123 @@
+
+C<key> is the an identifier for the configuration value.
+
+C<type> species the type of the value...
```
Are the `...`s in this POD suggesting more is yet to be written?

tmooney commented on lib/perl/Genome/Configurable.pm:
```diff
@@ -0,0 +1,69 @@
+
+sub set_prop_desc_with_warn {
+    my ($prop_desc, $key, $value) = @_;
+    if ($prop_desc->{$key}) {
```
A false value seems bogus, anyway, but would it better capture the intent if this were a `defined` or `exists` check instead?

tmooney commented:
I like the way this is turning out!  (And as a bonus `Genome::Site::TGI::LegacyConfig` might finally go away, too :smile:)

nnutter commented:
CI, test this.

tmooney commented on lib/perl/Genome/Config/Command/Get.pm:
```diff
@@ -0,0 +1,31 @@
+
+sub execute {
+    my $self = shift;
+    my @keys = $self->keys || Genome::Config::all_keys();
```
Using `$self->keys` as a logical value will mean `@keys` gets the count instead of the keys themselves.

nnutter replied:
Thanks.  I also discovered that the new `calculated_default` calls in scalar context. ;-)

nnutter replied:
Actually, I guess I'm just using it wrong.  It should probably return an arrayref.

tmooney commented:
Should we move the existing modules under `Genome/Config/` to maybe claim the `Genome/AnalysisProject/` namespace instead?  (Or, possibly other things go under `Genome/Config/AnalysisProject/` and we leave that alone?)  They're sort of related, but sort of not.  

(For example, this *could* use `Genome::Config::Parser` to parse the YAML, but that seems kind of coincidental.)

nnutter commented:
> Should we move the existing modules under `Genome/Config/` to maybe claim the `Genome/AnalysisProject/` namespace instead?  (Or, possibly other things go under `Genome/Config/AnalysisProject/` and we leave that alone?)  They're sort of related, but sort of not.  

We'll probably want to do something.  In my mind that can wait until towards the end of this PR or maybe even as a follow-up PR.

nnutter commented:
I did another rebase (mostly to pickup a UR fix).  aab374b is the equivalent of the pre-rebase state.  I've added commits that remove a couple environment variables and converted a couple configuration variables.

nnutter commented:
I'm don't think the `type` means much in the `ConfigSpec` so I'm going to rethink that a bit.  I think this gets back to one of the feedback @acoffman had.

I'm also debating exporting a `config()` or `get_config()` (or similar) to use instead of `Genome::Config::get()`.

nnutter commented:
To continue to try and keep the PR reviewable I'm separating out the conversion of the codebase as a separate PR, see nnutter/genome#1.

nnutter commented:
CI, test this.

nnutter commented:
I screwed up nnutter/genome#1 by accident so now that work is on nnutter/genome#2.

nnutter commented:
CI, test this.

tmooney commented on lib/perl/Genome/Test/Config.pm:
```diff
@@ -0,0 +1,81 @@
+sub temp_dir_helper {
+    my @temp_dirs;
+    my $new_temp_dir = sub {
+        push @temp_dirs, File::Temp->newdir();
```
It might be good if this had the behaviour of `Genome::Sys->create_temp_directory` where it is aware of LSF's temp directory location when `bsub`bed.

tmooney commented on lib/perl/Genome/Config.t:
```diff
@@ -0,0 +1,137 @@
+subtest 'basic lookup' => sub {
+    plan tests => 2;
+
+    my ($temp_dirs, $new_temp_dir) = Genome::Test::Config::temp_dir_helper();
```
I believe that, as long as the caller here holds a reference to `$new_temp_dir`, the `@temp_dirs` inside will not be cleaned up even if a reference to that array is not held here.

nnutter replied:
I think you are describing what I wanted.  I want this scope/`subtest` to hold the (only) reference to the `File::Temp` object so that it is cleaned up when the `subtest` ends.

nnutter commented:
One major decision I have left is whether to migrate everything to a new environment variable prefix to break ties with the legacy system.  The cost of that is missing any the places where we have override `/etc/profile.d/genome.sh`.

nnutter commented:
I'll need to coordinate changes to `puppet-modules` and `wugc-hudson` as part of this.

nnutter commented:
I've pulled out the `XGENOME_` prefix change and will do that as a separate PR.

nnutter commented:
I've removed the _work in progress_ label as I think all major changes are done.  I think this is close to being ready to ship.  I will be following this up PR up with a separate that changes the environment variables prefixes so we can get a clean break from the legacy system.

nnutter commented:
If it helps anyone wanting to help review, right now the newly added files are:

```
etc/genome/*
lib/perl/Genome/Config.t
lib/perl/Genome/Config/Command/Get.pm
lib/perl/Genome/Config/Command/Get.t
lib/perl/Genome/Config/Command/List.pm
lib/perl/Genome/Config/Command/List.t
lib/perl/Genome/Config/Command/SetEnv.pm
lib/perl/Genome/Config/Command/SetEnv.t
lib/perl/Genome/Config/Command/Validate.pm
lib/perl/Genome/Config/Command/Validate.t
lib/perl/Genome/ConfigSpec.pm
lib/perl/Genome/ConfigSpec.t
lib/perl/Genome/ConfigValidator/Defined.pm
lib/perl/Genome/ConfigValidator/LSFQueue.pm
lib/perl/Genome/ConfigValidator/Numeric.pm
lib/perl/Genome/ConfigValidator/Numeric.t
lib/perl/Genome/ConfigValidator/Port.pm
lib/perl/Genome/ConfigValidator/Positive.pm
lib/perl/Genome/ConfigValidator/UR.pm
lib/perl/Genome/ConfigValidator/URI.pm
lib/perl/Genome/ConfigValidator/URL.pm
lib/perl/Genome/ConfigValidatorBase.pm
lib/perl/Genome/Configurable.pm
lib/perl/Genome/Configurable.t
lib/perl/Genome/Test/Config.pm
```

Found by doing `git log --name-status @{u}.. | grep '^A       ' | awk '{ print $2 }' | sort -u`.

You should definitely look at `Genome::Config` to see how values are resolved.  Basically it's environment > home > snapshot > global.   For commands we'll be able to use `Genome::Configurable` which will allow the default value for an option/parameter to be looked up that way so for `Genome::Configurable` it's command line > environment > home > snapshot > global.  Shell scripts can make use of `genome config get` and `genome config set-env` to get or set values.

nnutter commented:
The `apipe-ci/bats` tests always run against `master` so fail.  I'll either have to ignore them or update them to run using the PR's genome/ur/workflow/sqitch code.

nnutter commented:
I've been waiting to merge this until model tests are passing again.  Once they are I'll merge it.
